docs/source/developer/ui_components.rst

```diff
@@ -0,0 +1,337 @@
+background image.
+
+Post jlab-2.0, nearly all icons in core are now created using
+`LabIcon <https://github.com/jupyterlab/jupyterlab/blob/f0153e0258b32674c9aec106383ddf7b618cebab/packages/ui-components/src/icon/labicon.tsx>`__
+and the icons-as-inline-svg pattern:
+
+-  Construct a new instance of LabIcon from the icon’s name and svg:
+
+   .. code:: typescript
+
+      // typescript
+
+      // svgstr is the raw contents of an icon's svg file
+      export const fooIcon = new LabIcon({
+        name: 'barpkg:foo',
+        svgstr: '<svg>...</svg>'
+      });
+
+-  Add the icon to the DOM using the appropriate property of your
+   LabIcon instance (either LabIcon.element() to directly create a DOM
+   node, or LabIcon.react to get the icon as a react component):
+
+   .. code:: typescript
+
+      // typescript
+
+      const e = fooIcon.element();
+      document.body.append(e);
+
+What you end up with is a DOM node (by default a ‘div’) that has an
+inline svg node as a child.
+
+``background-image`` vs inline svg
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+The big limitation of the old icon-as-css-background pattern is that svg
+images rendered as ``background-image`` are invisible to CSS. On the
+other hand, an icon rendered as an inline svg node is fully exposed to
+the CSS. This allows us to dynamicly change icon styling as needed
+simply by modifying our CSS. Most importantly, this allows us to recolor
+icons according to Jupyterlab’s current theme.
+
+How JupyterLab handles icons
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The @jupyterlab/ui-components package provides icons to the rest of
+JupyterLab, in the form of a set of ``LabIcon`` instances (currently
+about 80). All of the icons in the core JupyterLab packages are rendered
+using one of these ``LabIcon`` instances.
+
+Using the icons in your own code
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+You can use any of JupyterLab icons in your own code via an ``import``
+statement. For example, to use ``jupyterIcon`` you would first do:
+
+.. code:: typescript
+
+   import { jupyterIcon } from '@jupyterlab/ui-components';
+
+How to render an icon into a DOM node
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Icons can be added as children to any ``div`` or ``span`` nodes using
+the ``icon.element(...)`` method (where ``icon`` is any instance of
+``LabIcon``). For example, to render the Jupyter icon you could do:
+
+.. code:: typescript
+
+   jupyterIcon.element({
+     container: elem,
+     height: '16px',
+     width: '16px',
+     marginLeft: '2px'
+   });
+
+where ``elem`` is any ``HTMLElement`` with a ``div`` or ``span`` tag. As
+shown in the above example, the icon can be styled by passing CSS
+parameters into ``.element(...)``. Any valid CSS parameter can be used
+(one catch: snake case params do have to be converted to camel case:
+instead of ``foo-bar: '8px'``, you’d need to use ``fooBar: '8px'``.
+
+How to render an icon as a React component
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

Comment:
I felt like there were more ways to use it that these two, but maybe I was mis-remembering? Like is there some way we use it by the "name"? Or what function does the name play?

Reply:
> Like is there some way we use it by the "name"?

Yes, there are the `LabIcon.resolveElement` and `LabIcon.resolveReact` static methods. The use case for these is very narrow; so far the only use case I've actually seen out in extensionland is for reusing our FileType icons in custom browser implementations. There's also a bunch of ways to shoot yourself in the foot with these. For now, I'm comfortable leaving the `resolveX` methods in the [API docs](http://jupyterlab.github.io/jupyterlab/ui-components/classes/labicon.html#resolveelement), where they are well covered...

at least that's what I would have liked to have said. I just now took a closer look at the API docs, and it looks like most of my docstrings are getting butchered by typedoc. Whelp, guess that's another thing to put on the list of "shit to fix in the API docs"

Reply:
OK! So are all the other methods/attributes (like `svgstr`) effectively deprecated?

Reply:
? I don't quite follow your logic there in declaring `svgstr` deprecated. There are a handful of public fields (`name`, `svgstr`, `render`) in a LabIcon instance that exist for the purpose of exposing metadata and/or integrating with external projects, eg `render` is used in the Lumino integration. `name` and `svgstr` are also exposed for the purpose of consistency across icon interfaces, so that eg `IFileType.icon` in `rendermime-interfaces` can take either a `LabIcon` instance or a simple `{name, svgstr}` pair. They aren't relevant for the extension dev docs, though I may add some more info about them to the API docs.